src/openrct2-ui/scripting/CustomWindow.cpp

```diff
@@ -0,0 +1,793 @@
+/*****************************************************************************
+ * Copyright (c) 2014-2018 OpenRCT2 developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/OpenRCT2/OpenRCT2
+ *
+ * OpenRCT2 is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#ifdef __ENABLE_SCRIPTING__
+
+#    include "../interface/Dropdown.h"
+#    include "ScUi.hpp"
+#    include "ScWindow.hpp"
+
+#    include <limits>
+#    include <openrct2-ui/interface/Widget.h>
+#    include <openrct2-ui/windows/Window.h>
+#    include <openrct2/drawing/Drawing.h>
+#    include <openrct2/localisation/Language.h>
+#    include <openrct2/localisation/Localisation.h>
+#    include <openrct2/localisation/StringIds.h>
+#    include <openrct2/scripting/Plugin.h>
+#    include <openrct2/world/Sprite.h>
+#    include <optional>
+#    include <string>
+#    include <vector>
+
+using namespace OpenRCT2;
+using namespace OpenRCT2::Scripting;
+
+namespace OpenRCT2::Ui::Windows
+{
+    enum
+    {
+        WIDX_BACKGROUND,
+        WIDX_TITLE,
+        WIDX_CLOSE,
+        WIDX_CONTENT_PANEL,
+        WIDX_CUSTOM_BEGIN,
+    };
+
+    static rct_widget CustomDefaultWidgets[] = {
+        { WWT_FRAME, 0, 0, 0, 0, 0, 0xFFFFFFFF, STR_NONE },                  // panel / background
+        { WWT_CAPTION, 0, 1, 0, 1, 14, STR_STRING, STR_WINDOW_TITLE_TIP },   // title bar
+        { WWT_CLOSEBOX, 0, 0, 0, 2, 13, STR_CLOSE_X, STR_CLOSE_WINDOW_TIP }, // close x button
+        { WWT_RESIZE, 1, 0, 0, 14, 0, 0xFFFFFFFF, STR_NONE },                // content panel
+    };
+
+    static void window_custom_close(rct_window* w);
+    static void window_custom_mouseup(rct_window* w, rct_widgetindex widgetIndex);
+    static void window_custom_mousedown(rct_window* w, rct_widgetindex widgetIndex, rct_widget* widget);
+    static void window_custom_resize(rct_window* w);
+    static void window_custom_dropdown(rct_window* w, rct_widgetindex widgetIndex, int32_t dropdownIndex);
+    static void window_custom_invalidate(rct_window* w);
+    static void window_custom_paint(rct_window* w, rct_drawpixelinfo* dpi);
+    static void window_custom_update_viewport(rct_window* w);
```

Comment:
Question: are we certain these callbacks will never be invoked with a `nullptr`? Because none of them is checking for this.

Reply:
They shouldn't be nullptr, no events will check for this.

Reply:
Ok